File: reports/safeforge_cosynthesized_decoder/reference/reference_model.cpp
// Generated bit-exact reference for safeforge-robust-8-4-v1-safe.
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <string>

static constexpr unsigned K = 4;
static constexpr unsigned R = 4;
static constexpr unsigned N = 8;
static constexpr std::uint64_t H_COLUMNS[N] = {10ULL, 5ULL, 12ULL, 7ULL, 1ULL, 2ULL, 4ULL, 8ULL};
static constexpr std::uint64_t G_ROWS[K] = {161ULL, 82ULL, 196ULL, 120ULL};

std::uint64_t syndrome(std::uint64_t word) {
    std::uint64_t value = 0;
    for (unsigned position = 0; position < N; ++position) if ((word >> position) & 1ULL) value ^= H_COLUMNS[position];
    return value;
}

std::uint64_t encode(std::uint64_t data) {
    std::uint64_t out = 0;
    for (unsigned source = 0; source < K; ++source) if ((data >> source) & 1ULL) out ^= G_ROWS[source];
    return out;
}

int main(int argc, char** argv) {
    if (argc != 3) return 2;
    const std::uint64_t data = std::stoull(argv[1], nullptr, 0);
    const std::uint64_t error = std::stoull(argv[2], nullptr, 0);
    const std::uint64_t codeword = encode(data);
    const std::uint64_t received = codeword ^ error;
    const std::uint64_t syn = syndrome(received);
    std::uint64_t mask = 0;
    bool known = false;
    switch (syn) {
        case 1ULL: mask = 16ULL; known = true; break;
        case 2ULL: mask = 32ULL; known = true; break;
        case 5ULL: mask = 2ULL; known = true; break;
        case 8ULL: mask = 128ULL; known = true; break;
        case 9ULL: mask = 6ULL; known = true; break;
        case 11ULL: mask = 12ULL; known = true; break;
        case 13ULL: mask = 9ULL; known = true; break;
        case 15ULL: mask = 3ULL; known = true; break;
        default: break;
    }
    std::string outcome;
    long long decoded = -1;
    if (syn == 0) {
        decoded = static_cast<long long>(received & ((1ULL << K) - 1ULL));
        outcome = decoded == static_cast<long long>(data) ? "correct" : "silent_corruption";
    } else if (!known) {
        outcome = "detected_uncorrectable";
    } else {
        const std::uint64_t corrected = received ^ mask;
        if (syndrome(corrected) != 0) outcome = "decoder_failure";
        else {
            decoded = static_cast<long long>(corrected & ((1ULL << K) - 1ULL));
            outcome = decoded == static_cast<long long>(data) ? "corrected" : "silent_corruption";
        }
    }
    std::cout << codeword << ' ' << received << ' ' << syn << ' ' << outcome << ' ' << decoded << '\n';
    return 0;
}

File: reports/safe_decoder_wasserstein/reference/reference_model.py
#!/usr/bin/env python3
"""Generated bit-exact reference for forge-hotspot-8-4-v1-safe."""
from __future__ import annotations
import argparse

K = 4
R = 4
N = 8
H_COLUMNS = [11, 14, 7, 13, 1, 2, 4, 8]
G = [[1, 0, 0, 0, 1, 1, 0, 1], [0, 1, 0, 0, 0, 1, 1, 1], [0, 0, 1, 0, 1, 1, 1, 0], [0, 0, 0, 1, 1, 0, 1, 1]]
DECODER = {1: 16, 3: 48, 5: 3, 8: 128, 9: 6, 10: 12, 15: 132}

def syndrome(word: int) -> int:
    value = 0
    for position, column in enumerate(H_COLUMNS):
        if (word >> position) & 1:
            value ^= column
    return value

def encode(data: int) -> int:
    if data < 0 or data >= (1 << K):
        raise ValueError("data does not fit")
    out = 0
    for bit in range(N):
        value = 0
        for source in range(K):
            value ^= ((data >> source) & 1) & G[source][bit]
        out |= value << bit
    return out

def decode(received: int, original_data: int | None = None) -> dict:
    syn = syndrome(received)
    if syn == 0:
        decoded = received & ((1 << K) - 1)
        outcome = "correct" if original_data is None or decoded == original_data else "silent_corruption"
        return {"outcome": outcome, "syndrome": syn, "decoded_data": decoded, "correction_mask": 0}
    if syn not in DECODER:
        return {"outcome": "detected_uncorrectable", "syndrome": syn, "decoded_data": None, "correction_mask": 0}
    mask = DECODER[syn]
    corrected = received ^ mask
    if syndrome(corrected) != 0:
        return {"outcome": "decoder_failure", "syndrome": syn, "decoded_data": None, "correction_mask": mask}
    decoded = corrected & ((1 << K) - 1)
    outcome = "corrected" if original_data is None or decoded == original_data else "silent_corruption"
    return {"outcome": outcome, "syndrome": syn, "decoded_data": decoded, "correction_mask": mask}

def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("data", type=lambda value: int(value, 0))
    parser.add_argument("error_mask", type=lambda value: int(value, 0))
    args = parser.parse_args()
    codeword = encode(args.data)
    result = decode(codeword ^ args.error_mask, args.data)
    print(f"{codeword} {codeword ^ args.error_mask} {result['syndrome']} {result['outcome']} {result['decoded_data'] if result['decoded_data'] is not None else -1}")

if __name__ == "__main__":
    main()
